sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -2138,199 +2287,28 @@ case class Cast(
   final override def nodePatternsInternal(): Seq[TreePattern] = Seq(CAST)
 
   override def canCast(from: DataType, to: DataType): Boolean = if (ansiEnabled) {
-    AnsiCast.canCast(from, to)
+    Cast.canAnsiCast(from, to)
   } else {
     Cast.canCast(from, to)
   }
 
+  private val (fallbackConfKey, fallbackConfValue) =
+    if (getTagValue(Cast.TABLE_INSERTION_RESOLVER).getOrElse(false)) {
```

Comment:
can we pass a boolean flag (existence of this tag) to `typeCheckFailureMessage` and determine the config names there?